subprojects/platform-jvm/src/integTest/groovy/org/gradle/jvm/toolchain/JavaToolchainDownloadSpiIntegrationTest.groovy

```diff
@@ -0,0 +1,376 @@
+        """
+
+        buildFile << """
+            apply plugin: "java"
+
+            java {
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(99)
+                }
+            }
+        """
+
+        file("src/main/java/Foo.java") << "public class Foo {}"
+
+        when:
+        failure = executer
+                .withTasks("compileJava")
+                .requireOwnGradleUserHomeDir()
+                .withToolchainDownloadEnabled()
+                .runWithFailure()
+
+        then:
+        failure.assertHasCause("Failed to apply plugin class 'UselessToolchainRegistry2Plugin'.")
+                .assertHasCause("Duplicate JavaToolchainRepository registration under the name 'uselessRegistry'")
+    }
+
+    @ToBeFixedForConfigurationCache(because = "Fails the build with an additional error")
+    def "list of requested repositories can be queried"() {
+        settingsFile << """
+            ${applyToolchainManagementBasePlugin()}
+            ${applyToolchainRegistryPlugin("uselessRegistry1", "UselessToolchainRegistry1", uselessToolchainRegistryCode("UselessToolchainRegistry1"))}            
+            ${applyToolchainRegistryPlugin("uselessRegistry2", "UselessToolchainRegistry2", uselessToolchainRegistryCode("UselessToolchainRegistry2"))}            
+            ${applyToolchainRegistryPlugin("uselessRegistry3", "UselessToolchainRegistry3", uselessToolchainRegistryCode("UselessToolchainRegistry3"))}            
+            toolchainManagement {
+                jdks {
+                    add("uselessRegistry3")
+                    add("uselessRegistry1")
+                }
+            }
+            
+            println(\"\"\"Explicitly requested toolchains: \${toolchainManagement.jdks.getAll().collect { it.getName() }}\"\"\")
+        """
+
+        buildFile << """
+            apply plugin: "java"
+
+            java {
+                toolchain {
+                    languageVersion = JavaLanguageVersion.of(99)
+                }
+            }
+        """
+
+        file("src/main/java/Foo.java") << "public class Foo {}"
+
+        when:
+        failure = executer
+                .withTasks("compileJava")
+                .requireOwnGradleUserHomeDir()
+                .withToolchainDownloadEnabled()
+                .runWithFailure()
+
+        then:
+        failure.getOutput().contains("Explicitly requested toolchains: [uselessRegistry3, uselessRegistry1]")
+    }
+
+    private static String applyToolchainManagementBasePlugin() {
+        //TODO (#21082): the base plugin which injects the "jdks" block will need to be a Gradle core plugin, this is an intermediate state of development
+        """
+            import org.gradle.internal.event.ListenerManager;
+            import org.gradle.jvm.toolchain.internal.DefaultJdksBlockForToolchainManagement;
+
+            public abstract class ToolchainManagementBasePlugin implements Plugin<Settings> {
+                @Inject
```

Comment:
Are these `@Inject` annotations required for `abstract` methods?

Reply:
Yes.